lib/cli/RenderAction/models/CacheManager.ts

```diff
@@ -0,0 +1,152 @@
+import type { RenderActionOptions } from "@/types";
+
+import type {
+  CacheEntry,
+  CacheEntryWithRenderOutput,
+} from "../utils/cache/types";
+
+import CacheEncryption from "./CacheEncryption";
+import CacheSerializer from "./CacheSerializer";
+import CacheBunFs from "./CacheBunFs";
+import CacheBunSqlite3 from "./CacheSqlite3";
+import RenderCacheHeaderDatabase from "./CacheHeaderDatabase";
+
+type CacheType = Exclude<RenderActionOptions["cacheType"], undefined>;
+
+type CacheManagerConstructorOptions = {
+  type: CacheType;
+  key: string;
+  component: {
+    name: string;
+    path: string;
+    props: unknown;
+  };
+};
+
+export default class CacheManager {
+  private static _hcache_db: RenderCacheHeaderDatabase =
+    new RenderCacheHeaderDatabase();
+
+  private id: string;
+  private type: CacheType;
+  private cpath: string;
+  private cname: string;
+  private cprops: unknown;
+  private cacheKey: string;
+
+  public static isInCache(cpath: string, cname: string, props: unknown) {
+    /**
+     * isInCache should not have to be o(n) because in such a case,
+     * we should just call find() and perform a single operation.
+     * However, if isInCache can be o(1) then it's a negligible check.
+     *
+     * Use a Map that cacheManagers have shared access to (Some singleton context)
+     * to store a hash of the cname,cpath,cprops
+     * convert cprops to deterministic structure
+     */
+    if (!CacheSerializer.serializable(props)) return false;
+    if (typeof props !== "object") return false;
+
+    return Boolean(
+      CacheManager._hcache_db.find({
+        cname: cname,
+        cpath: cpath,
+        props: props,
+        id: CacheManager.createHeaderCacheKey(cpath, cname),
+      }),
+    );
+  }
+
+  private static createHeaderCacheKey(cpath: string, cname: string) {
+    return CacheEncryption.sha256Hash(cpath.concat("::", cname));
+  }
+
+  private static addToHeaderCache(
+    cpath: string,
+    cname: string,
+    props: unknown,
+  ): boolean {
+    try {
+      return CacheManager._hcache_db.add({
+        cname,
+        cpath,
+        props,
+        id: CacheManager.createHeaderCacheKey(cpath, cname),
```

Comment:
We're gonna get duped IDs like this. You need to cache the serialized props into the ID as well